quil-rs/src/stats/mod.rs

```diff
@@ -0,0 +1,270 @@
+mod execution_graph;
+
+#[cfg(test)]
+pub(crate) mod test_programs;
+
+// Use quil_rs::program::graph for pulse-level programs, but not qubit programs.
+
+use std::collections::HashSet;
+use std::ops::Neg;
+
+use crate::{
+    instruction::{Instruction, Qubit},
+    Program,
+};
+use execution_graph::{Error as ExecutionGraphError, ExecutionGraph};
+
+pub trait InstructionsSource {
+    fn body_instructions(&self) -> impl Iterator<Item = &Instruction> + '_;
+    fn num_instructions(&self) -> usize;
+    fn get_used_qubits(&self) -> impl IntoIterator<Item = &Qubit>;
+}
+
+impl InstructionsSource for Program {
+    fn body_instructions(&self) -> impl Iterator<Item = &Instruction> + '_ {
+        Program::body_instructions(self)
+    }
+
+    fn num_instructions(&self) -> usize {
+        Program::num_instructions(self)
+    }
+
+    // There's no generic trait for "set-like" types, so we return a concrete
+    // type here (as an API convenience) but do not require this type in the
+    // trait definition.
+    #[allow(refining_impl_trait)]
+    fn get_used_qubits(&self) -> &HashSet<Qubit> {
+        Program::get_used_qubits(self)
+    }
+}
+
+impl InstructionsSource for &Program {
+    fn body_instructions(&self) -> impl Iterator<Item = &Instruction> + '_ {
+        (*self).body_instructions()
+    }
+
+    fn num_instructions(&self) -> usize {
+        (*self).num_instructions()
+    }
+
+    // See note on `Program` impl above.
+    #[allow(refining_impl_trait)]
+    fn get_used_qubits(&self) -> &HashSet<Qubit> {
+        (*self).get_used_qubits()
+    }
+}
+
+pub struct ProgramStats<S: InstructionsSource> {
+    source: S,
+    execution_graph: ExecutionGraph,
+}
+
+fn make_execution_graph<S: InstructionsSource>(
+    source: S,
+) -> Result<ExecutionGraph, ExecutionGraphError> {
+    ExecutionGraph::new(source.body_instructions().cloned())
+}
+
+impl<'a> ProgramStats<&'a Program> {
+    pub fn from_program(program: &'a Program) -> Result<Self, ExecutionGraphError> {
+        let execution_graph = make_execution_graph(program)?;
+
+        Ok(Self {
+            source: program,
+            execution_graph,
+        })
+    }
+
+    // Provided because Program specifically returns a hash-set, but in general
+    // `InstructionsSource` may not provide that feature.
+    pub fn set_of_qubits_used(&self) -> &HashSet<Qubit> {
+        self.source.get_used_qubits()
+    }
```

Comment:
If we used associated types for the trait, this would be unnecessary, since the actual type would be exposed as a `HashSet`.